borsh/tests/test_recursive_structs.rs

```diff
@@ -0,0 +1,66 @@
+#![cfg(feature = "derive")]
+use borsh::{from_slice, BorshDeserialize, BorshSerialize};
+
+#[cfg(feature = "hashbrown")]
+use hashbrown::HashMap;
+
+#[cfg(feature = "std")]
+use std::collections::HashMap;
+
+#[cfg(hash_collections)]
+use core::{cmp::Eq, hash::Hash};
+
+#[cfg(not(feature = "std"))]
+extern crate alloc;
+#[cfg(not(feature = "std"))]
+use alloc::{boxed::Box, string::String, vec::Vec};
+
+#[cfg(hash_collections)]
+#[derive(BorshSerialize, BorshDeserialize)]
+struct CRec<U: PartialOrd + Hash + Eq> {
+    a: String,
+    b: HashMap<U, CRec<U>>,
+}
+
+//  `impl<T, U> BorshDeserialize for Box<T>` pulls in => `ToOwned`
+// => pulls in at least `Clone`
```

Comment:
`Clone` isn't actually used, it's needed for associated [`ToOwned` type bound](https://doc.rust-lang.org/src/alloc/borrow.rs.html#86)
Serde does it simpler:

```rust
impl<T, U> BorshDeserialize for Box<T>
where
    U: Into<Box<T>> + Borrow<T>,
    T: ToOwned<Owned = U> + ?Sized,
    T::Owned: BorshDeserialize,
{
    fn deserialize_reader<R: Read>(reader: &mut R) -> Result<Self> {
        Ok(T::Owned::deserialize_reader(reader)?.into())
    }
}
```

```rust
impl<'de, T: Deserialize<'de>> Deserialize<'de> for Box<T> {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        Deserialize::deserialize(deserializer).map(Box::new)
    }
}
```